src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -2364,44 +2363,56 @@ void processResource(IOperation resource, ArrayBuilder<(IVariableDeclarationOper
                 Debug.Assert(_currentRegion.Kind == ControlFlowGraph.RegionKind.Try);
                 LeaveRegion();
 
-                var endOfFinally = new BasicBlock(BasicBlockKind.Block);
-                endOfFinally.InternalNext.Branch.Kind = BasicBlock.BranchKind.StructuredExceptionHandling;
+                AddDisposingFinally(resource, knownToImplementIDisposable: true, iDisposable, compilation);
 
-                EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.Finally));
-                AppendNewBlock(new BasicBlock(BasicBlockKind.Block));
+                Debug.Assert(_currentRegion.Kind == ControlFlowGraph.RegionKind.TryAndFinally);
+                LeaveRegion();
 
-                if (!(resource.Type?.IsValueType == true && !ITypeSymbolHelpers.IsNullableType(resource.Type)))
-                {
-                    IOperation condition = MakeIsNullOperation(OperationCloner.CloneOperation(resource), compilation);
-                    condition = Operation.SetParentOperation(condition, null);
-                    LinkBlocks(CurrentBasicBlock, (condition, JumpIfTrue: true, RegularBranch(endOfFinally)));
-                    _currentBasicBlock = null;
-                }
+                AppendNewBlock(afterTryFinally, linkToPrevious: false);
+            }
+        }
 
-                if (!resource.Type.Equals(iDisposable))
-                {
-                    resource = convertToIDisposable(resource);
-                }
+        private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisposable, ITypeSymbol iDisposable, Compilation compilation)
+        {
+            Debug.Assert(_currentRegion.Kind == ControlFlowGraph.RegionKind.TryAndFinally);
 
-                AddStatement(tryDispose(resource) ??
-                             // PROTOTYPE(dataflow): The scenario with missing Dispose is not covered by unit-tests.
-                             MakeInvalidOperation(type: null, resource));
+            var endOfFinally = new BasicBlock(BasicBlockKind.Block);
+            endOfFinally.InternalNext.Branch.Kind = BasicBlock.BranchKind.StructuredExceptionHandling;
 
-                AppendNewBlock(endOfFinally);
+            EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.Finally));
+            AppendNewBlock(new BasicBlock(BasicBlockKind.Block));
 
-                LeaveRegion();
-                Debug.Assert(_currentRegion.Kind == ControlFlowGraph.RegionKind.TryAndFinally);
-                LeaveRegion();
+            if (!knownToImplementIDisposable)
+            {
+                Debug.Assert(!(resource.Type?.IsValueType == true && !ITypeSymbolHelpers.IsNullableType(resource.Type)));
```

Comment:
All the inversions in this assert make it very complicated to read and reason about simply. Please simplify it.

Reply:
Or consider making a helper, as this condition is used multiple times.

---
In reply to: [181883394](https://github.com/dotnet/roslyn/pull/26158#discussion_r181883394) [](ancestors = 181883394)

Reply:
> All the inversions in this assert make it very complicated to read and reason about simply. Please simplify it.

Inversion is not trivial when dealing with nullable values. 

---
In reply to: [181883394](https://github.com/dotnet/roslyn/pull/26158#discussion_r181883394) [](ancestors = 181883394)